### ISSUE TITLE
Counter-intuitive Patience & Cooldown of ReduceLROnPlateau

### ISSUE THREAD

**Issue by Jiaming-Liu** [pytorch/pytorch#2545]
https://github.com/pytorch/pytorch/pull/1370 provides a LR scheduler that reduces LR when the validation metrics stops improving. 

It works like this: when ```patience=2, cooldown=0```,
``` Python 
import torch
from torch.optim.lr_scheduler import ReduceLROnPlateau
n=torch.nn.Linear(1,1)
opt = torch.optim.SGD(n.parameters(), lr=0.1)
rp = ReduceLROnPlateau(opt, patience=2, cooldown=0)
for _ in range(10):
    rp.step(metrics=0) # metrics "stops improving" after the first epoch
    print(opt.param_groups[0]['lr'])
```

Output:
```
0.1
0.1
0.1
0.01
0.01
0.01
0.001
0.001
0.001
0.0001
```

when `patience=2, cooldown=1`,
```
0.1
0.1
0.1
0.01
0.01
0.01
0.01
0.001
0.001
0.001
```

IMO the two examples should be matched to `patience=2, cooldown=1` and `patience=2, cooldown=2` respectively. Any idea?

**Comment by apaszke** [comment]
Yea it seems there's a bug somewhere

**Comment by Jiaming-Liu** [comment]
I guess we simply need to change `if self.num_bad_epochs > self.patience` to `>=`.
Also, a note for myself, we should check the type & value of `patience`.

**Comment by Jiaming-Liu** [comment]
Another test case:
``` Python
import torch
from torch.optim.lr_scheduler import ReduceLROnPlateau
n=torch.nn.Linear(1,1)
opt = torch.optim.SGD(n.parameters(), lr=0.1)
rp = ReduceLROnPlateau(opt, patience=2, cooldown=0)
metrics = [10.,10.,10.,1.,1.,1.,0.1,0.1,0.1,0.1,0.1]
for epoch in range(10):
    print('---------')
    print('epoch: %d' % epoch)
    print(opt.param_groups[0]['lr'])
    # train_one_epoch(...)
    # metrics = validate(...)
    print('metrics: %f' % metrics[epoch])
    rp.step(metrics=metrics[epoch])
```

Result:
```
---------
epoch: 0
0.1
metrics: 10.000000
'UPDATED_BEST'
---------
epoch: 1
0.1
metrics: 10.000000
'WAITING'
---------
epoch: 2
0.1
metrics: 10.000000
'WAITING'
---------
epoch: 3
0.1
metrics: 1.000000
'UPDATED_BEST'
---------
epoch: 4
0.1
metrics: 1.000000
'WAITING'
---------
epoch: 5
0.1
metrics: 1.000000
'WAITING'
---------
epoch: 6
0.1
metrics: 0.100000
'UPDATED_BEST'
---------
epoch: 7
0.1
metrics: 0.100000
'WAITING'
---------
epoch: 8
0.1
metrics: 0.100000
'WAITING'
---------
epoch: 9
0.1
metrics: 0.100000
'REDUCED_LR'
```
